### PR TITLE
[go-server] Allow user to specify how errors are handled

### PR DESCRIPTION
**Pull request by lwj5**
- Added `error.go` with an error handler
- Updated controller template to use the error handler
- Backwards compatible with previous version

Tested on both go-server with `mux` and `chi` routers

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

Hi @antihax @grokify @kemokemo @jirikuncar @ph4r5h4d, please help to take a look at this PR

### CODE REVIEW

**Comment by lwj5** [comment]
`continuous-integration/appveyor/pr` looks like a c sharp error and is unrelated

**Comment by ph4r5h4d** [comment]
I'd like to see more feedback on this PR. I'm not getting what the error handler is trying to solve or improve.

**Comment by lwj5** [comment]
Let's assume there are 2 types of users of the generator:

1. those that edit the generated files, and
2. those that extend (or embed) the generated files.

#### For type 1 users

If their specifications require them to return errors with fields like `title` and `details`. You would need to replace all instances of `w.WriteHeader(http.StatusBadRequest)` with your own. This is tedious and will require you to exclude the file from further regeneration if there's a spec change as it will be overwritten. If the concern of error handling is separated from the controller, it would mean that most (if not all) of the time, the controller can be left as-is.

- Modifying `error.go` provides a way convenient to customize the error responses without changing the controller, separating their concerns.

Second, let's say you have a debug mode in your application. If debug is turned on, the details of the error responses are more verbose, and if off, it merely returns "unexpected error". `error.go` provides a convenient place to process the errors returned from the servicer before the response is sent.

- `error.go` provides a centralized place to process the errors returned from the servicer.

#### Case 2 is where it shines

Where I work, we do not modify any of the generated files. Instead, we use overlays that embed the servicer. If we have such a spec as laid out in the first point, we will need to reimplement the controller. In addition to the tediousness, this process is extremely error-prone especially if the specs are changing. The concerns could be cleanly separated with an injection of an `ErrorHandler`.

- Instead of reimplementing the controller, an appropriate `ErrorHandler` can be injected into the controller together with the servicer `NewController(s, errorHandlerOpt)`

To sum it up, `ErrorHandler` has a very specific job of handling error, like how `controller` and `servicer` has very specific roles.

Other than these, `EncodeJSONResponse(err.Error(), &result.Code, result.Headers, w)` is quite concerning as most errors are plain text. But with this it's easier to swap it out in the next PR.

**Comment by ph4r5h4d** [comment]
Thank you @lwj5, Now I get a clearer picture of what you had in mind when you submitted this PR.
I'll try my best to test it in my local and verify the functionality as fast as possible (hence I'm a bit busy these days and sorry for that).

P.S: In the meantime, I'd like to know others' opinions regarding this change.

**Comment by lwj5** [comment]
No worries, thanks for doing the review

**Comment by wing328** [comment]
@lwj5 thanks for the detailed explanation. One minor suggestion is to add some comments in https://github.com/OpenAPITools/openapi-generator/pull/9764/files#diff-40523ac4e3f7daf99f002d095c93bb305acc3a313f05b8160654207a3661a942 explaining what those functions do so as to onboard new Go users more easily.

**Comment by lwj5** [comment]
I rebased it with the master, the last 3 commits are new. 

- Added docs
- Remove variadic leads to cleaner code
- Regenerated files

**Comment by lwj5** [comment]
I've been using this branch with a custom error handler for a bit, it is working as intended and I've managed to inject debug and stack trace information to err responses originating from the servicers. This is all done without the controller/servicer needing to know whether the debug flag is set.

**Comment by lwj5** [comment]
@wing328 @ph4r5h4d any updates on this? I would like this merged so I can continue working on #9647 to meet the goal of #426

**Comment by ph4r5h4d** [comment]
> @wing328 @ph4r5h4d any updates on this? I would like this merged so I can continue working on #9647 to further the goal of #426 and hopefully to meet #1931, #3547

I'm sorry for the delay. It's been a hectic couple of weeks for me. I'll plan to test it today or at most tomorrow.

File: modules/openapi-generator/src/main/resources/go-server/controller-api.mustache [hunk 1]
```diff
@@ -19,11 +19,31 @@ import (
 // A {{classname}}Controller binds http requests to an api service and writes the service results to the http response
 type {{classname}}Controller struct {
 	service {{classname}}Servicer
+	errorHandler ErrorHandler
+}
+
+// Option for how controller is set up.
```
Comment (ph4r5h4d) [inline modules/openapi-generator/src/main/resources/go-server/controller-api.mustache:25]: The following is suggested to comply with Golang comment standards.
```suggestion
// {{classname}} for how the controller is set up.
```

File: samples/server/petstore/go-api-server/go/error.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+// ErrorHandler defines the required method for handling error. You may implement it and inject this into a controller if
+// you would like errors to be handled differently from the DefaultErrorHandler
+type ErrorHandler func(w http.ResponseWriter, r *http.Request, err error, result *ImplResponse)
+
+// DefaultErrorHandler defines the default logic on how to handle errors from the controller. Any errors from parsing
+// request params will return a StatusBadRequest. Otherwise, the error code originating from the servicer will be used.
+func DefaultErrorHandler(w http.ResponseWriter, r *http.Request, err error, result *ImplResponse) {
+	if _, ok := err.(*ParsingError); ok {
+		// Handle parsing errors
+		EncodeJSONResponse(err.Error(), func(i int) *int { return &i }(http.StatusBadRequest), map[string][]string{}, w)
```
Comment (ph4r5h4d) [inline samples/server/petstore/go-api-server/go/error.go:38]: In `EncodeJSONResponse` function you are returning an error, it's not handled when it's used. I understand that the case might be uncommon to happen but I suggest the following:
```suggestion
		err := EncodeJSONResponse(err.Error(), func(i int) *int { return &i }(http.StatusBadRequest), w)
		if err != nil {
			return 
		}
```
you can log the error or just put a TODO for an end consumer to be aware of the thing that might require attention.

File: samples/server/petstore/go-api-server/go/error.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	} else {
```
Comment (ph4r5h4d) [inline samples/server/petstore/go-api-server/go/error.go:39]: Here If I'm not wrong you can ignore the else part, do we need it?

File: samples/server/petstore/go-api-server/go/error.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		// Handle all other errors
+		EncodeJSONResponse(err.Error(), &result.Code, result.Headers, w)
```
Comment (ph4r5h4d) [inline samples/server/petstore/go-api-server/go/error.go:41]: The same can be done like the if part.

File: samples/server/petstore/go-api-server/go/error.go (hunk 2 above)
Comment (lwj5) [reply samples/server/petstore/go-api-server/go/error.go:38]: This is how all errors are being handled by the controller prior to the PR. I'm okay with removing it, but it would mean a change in the implementation for those currently using the generator.

File: samples/server/petstore/go-api-server/go/error.go (hunk 4 above)
Comment (lwj5) [reply samples/server/petstore/go-api-server/go/error.go:41]: similar to above

File: samples/server/petstore/go-api-server/go/error.go (hunk 3 above)
Comment (lwj5) [reply samples/server/petstore/go-api-server/go/error.go:39]: similar to the above. I'm ok to removing it

File: samples/server/petstore/go-api-server/go/error.go (hunk 2 above)
Comment (ph4r5h4d) [reply samples/server/petstore/go-api-server/go/error.go:38]: Oh sorry i made the comments in the wrong file, I'm moving it to the correct one.

File: samples/server/petstore/go-api-server/go/error.go (hunk 3 above)
Comment (ph4r5h4d) [reply samples/server/petstore/go-api-server/go/error.go:39]: It's up to you then :)

**Comment by ph4r5h4d** [comment]
This PR LGTM, I've tested it locally, and it works fine.
I can approve it now @lwj5 if all the changes are in place.

And sorry for the delay :)

**Comment by lwj5** [comment]
Yes, changes completed. It's better to create a new PR for changing the default error handling.

**Comment by ph4r5h4d** [comment]
@lwj5 I think you need to resubmit the samples.
`Please run 'bin/generate-samples.sh' locally and commit changes (UNCOMMITTED CHANGES ERROR)`
One pipeline is failing, kindly run the commands locally and see if they are changes to submit.

**Comment by lwj5** [comment]
Let me run that now

**Comment by ph4r5h4d** [comment]
@wing328 LGTM

**Review by ph4r5h4d — APPROVED** [review]

**Comment by lwj5** [comment]
@wing328

**Comment by wing328** [comment]
CircleCI failure not related to this change.
